### PR TITLE
More H2 Queries

### LINKED ISSUE #101: Query metrics: Chapter 20. HTTP/2
Part | Chapter | Authors | Reviewers | Tracking Issue
-- | -- | -- | -- | --
IV. Content Distribution | 20. HTTP/2 | @bazzadp | @bagder @rmarx @dotjs | #22

# READ ME!
All of the metrics in the table below have been marked as `Able To Query` during the [metrics triage](https://github.com/HTTPArchive/almanac.httparchive.org/issues/33). The analyst assigned to each metric is expected to write the corresponding query and submit a PR to [have it reviewed and added to the repo](https://github.com/HTTPArchive/almanac.httparchive.org/issues/62).

In order to stay on schedule and have the data ready for authors, **please have all metrics reviewed and merged by August 5**.

## Assignments

ID | Metric description | Analyst | Notes
-- | -- | -- | --
20.01 | Adoption rate of HTTP/2 by site (home page only) and by requests (all request on page) over the years. Trend graph over all available years. | @paulcalvano |  
20.03 | Average percentage of resources loaded over HTTP/2 (or gQUIC) versus HTTP/1.1 per site. Trend graph over all available years. | @paulcalvano |  
20.04 | Number of HTTP (not HTTPS) sites which return upgrade HTTP header containing h2. Once off stat for last crawl. | @paulcalvano |  
20.05 | Number of HTTPS sites using HTTP/2 which return upgrade HTTP header containing h2. Once off stat for last crawl. | @paulcalvano |  
20.06 | Number of HTTPS sites not using HTTP/2 which return upgrade HTTP header containing h2. Once off stat for last crawl. | @paulcalvano |  
20.07 | % of sites affected by CDN prioritization issues (H2 and served by CDN) - https://github.com/andydavies/http2-prioritization-issues#cdns--cloud-hosting-services. Once off stat for last crawl. | @paulcalvano | Possible per Barry's comment
20.08 | Count of HTTP/2 sites grouped by server HTTP header value but strip version numbers (e.g. Apache and Apache 2.4.28 and Apache 2.4.29 should all report as Apache, but Apache Tomcat should report as Tomcat. Probably need to massive the results to achieve this). Once off stat for last crawl. | @paulcalvano |  
20.09 | Count of non-HTTP/2 sites grouped by server HTTP header value but strip version numbers. Once off stat for last crawl. | @paulcalvano |  
20.10 | Count of HTTP/2 sites which use HTTP/2 Push. Trend graph over all available years. | @paulcalvano | HAR files - "_was_pushed": 1,
20.11 | Average number of HTTP/2 Pushed resources and average bytes. Once off stat for last crawl. | @paulcalvano | HAR files - "_was_pushed": 1,
20.12 | Count and number of bytes pushed by asset type (CSS, JS, Images...etc.). Once off stat for last crawl. | @paulcalvano | HAR files - "_was_pushed": 1,
20.13 | Count of preload HTTP Headers with nopush attribute set. Once off stat for last crawl. | @paulcalvano |  
20.15 | Measure number of TCP Connections per site. Average number of domains per site still going down year on year as per HTTP Archive State of the Web report? Trend graph over all available years. | @paulcalvano |  
20.16 | Measure average number of TCP Connections per site for HTTP/1.1 sites versus HTTP/2 sites. Once off stat for last crawl. | @paulcalvano |  

## Checklist of metrics to be merged

- [x] 	20.01	Adoption rate of HTTP/2 by site (home page only) and by requests (all request on page) over the years. Trend graph over all available years.
- [x] 	20.03	Average percentage of resources loaded over HTTP/2 (or gQUIC) versus HTTP/1.1 per site. Trend graph over all available years.
- [x] 	20.04	Number of HTTP (not HTTPS) sites which return upgrade HTTP header containing h2. Once off stat for last crawl.
- [x] 	20.05	Number of HTTPS sites using HTTP/2 which return upgrade HTTP header containing h2. Once off stat for last crawl.
- [x] 	20.06	Number of HTTPS sites not using HTTP/2 which return upgrade HTTP header containing h2. Once off stat for last crawl.
- [x] 	20.07	% of sites affected by CDN prioritization issues (H2 and served by CDN) - https://github.com/andydavies/http2-prioritization-issues#cdns--cloud-hosting-services. Once off stat for last crawl.
- [x] 	20.08	Count of HTTP/2 sites grouped by server HTTP header value but strip version numbers (e.g. Apache and Apache 2.4.28 and Apache 2.4.29 should all report as Apache, but Apache Tomcat should report as Tomcat. Probably need to massive the results to achieve this). Once off stat for last crawl.
- [x] 	20.09	Count of non-HTTP/2 sites grouped by server HTTP header value but strip version numbers. Once off stat for last crawl.
- [x] 	20.10	Count of HTTP/2 sites which use HTTP/2 Push. Trend graph over all available years.
- [x] 	20.11	Average number of HTTP/2 Pushed resources and average bytes. Once off stat for last crawl.
- [x] 	20.12	Count and number of bytes pushed by asset type (CSS, JS, Images...etc.). Once off stat for last crawl.
- [x] 	20.13	Count of preload HTTP Headers with nopush attribute set. Once off stat for last crawl.
- [x] 	20.15	Measure number of TCP Connections per site. Average number of domains per site still going down year on year as per HTTP Archive State of the Web report? Trend graph over all available years.
- [x] 	20.16	Measure average number of TCP Connections per site for HTTP/1.1 sites versus HTTP/2 sites. Once off stat for last crawl.

### PR DESCRIPTION
**Pull request by paulcalvano**
Closes #101

### CODE REVIEW

File: sql/2019/20_HTTP_2/20_06.sql [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    });
+    return st['value'];
+  } catch (e) {
+    return '';
+  }
+""";
+
+SELECT 
+  _TABLE_SUFFIX AS client,
+  JSON_EXTRACT_SCALAR(payload, "$._is_base_page") AS is_base_page,
```
Comment (rviscomi) [inline sql/2019/20_HTTP_2/20_06.sql]: Ditto previous comments. Should `is_base_page` be a field in the output or a condition in the WHERE clause? Applies to the rest of these queries as well.

**Comment by paulcalvano** [comment]
Added 20.16, which includes alt-svc headers at @bazzadp's request.

File: sql/2019/20_HTTP_2/20_04.sql [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    return st['value'];
+  } catch (e) {
+    return '';
+  }
+""";
+
+SELECT 
+  _TABLE_SUFFIX AS client,
+  JSON_EXTRACT_SCALAR(payload, "$._is_base_page") AS is_base_page,  
+  COUNT(*) AS num_requests
```
[the remark this answers is not in the document]
Comment (paulcalvano) [reply sql/2019/20_HTTP_2/20_04.sql]: Seeing a discrepancy in the results between the query in my PR and the one you shared above.   In the screenshot below, the query on the left processes the  httparchive.requests tables and the one on the right processes the httparchive.almanac.requests table.   

The one on the left shows 1890586 mobile HTTP/2 pages.  The one on the right shows 1996270 mobile HTTP/2 pages . I don't see any reason for the discrepency in the queries.  Any ideas? 

image

[the remark this answers is not in the document]
Comment (rviscomi) [reply sql/2019/20_HTTP_2/20_04.sql]: Great catch. Seems like a difference in requests marked `_is_base_page` and `firstHtml`. The former is a WPT annotation while the latter is determined in the HA pipeline.

Looking at the number of base pages:

```sql
#standardSQL
SELECT
  _TABLE_SUFFIX AS client,
  COUNT(0)
FROM
  `httparchive.requests.2019_07_01_*`
WHERE
  JSON_EXTRACT_SCALAR(payload, '$._is_base_page') = 'true'
GROUP BY
  client
```

There are 5,297,105 mobile and 4,371,570 desktop base pages. Compared to the right side totals in your screenshot, which are 5,558,214 and 4,550,580 respectively.

The actual number of pages in the `summary_pages` dataset for the 2019_07 crawl is:

```sql
#standardSQL
SELECT
  _TABLE_SUFFIX AS client,
  COUNT(0)
FROM
  `httparchive.summary_pages.2019_07_01_*`
GROUP BY
  client
```

5,297,442 and 4,371,973 respectively. So much closer to the base page numbers. I'm not even sure how we get ~200k _more_ `firstHtml` requests than there are pages. 😖 

```sql
#standardSQL
SELECT
  client,
  page,
  COUNT(0) AS firstHtml
FROM
  `httparchive.almanac.requests`
WHERE
  firstHtml
GROUP BY
  client,
  page
HAVING
  firstHtml > 1
ORDER BY
  firstHtml DESC
```

This reveals MANY pages in which there are more than 1 `firstHtml` value. Some with 400+. On [this example page](http://httparchive.webpagetest.org/result/190713_Dx89_G5WS/1/details/#waterfall_view_step1) the page makes over 1000 requests, many to `/index.php` and `/`. Maybe the HA pipeline saw the first request for `/`, marked it as `firstHtml`, then marked all subsequent requests for that URL also as `firstHtml`?

In any case, this bug affects all other queries that depend on `firstHtml`. I think the best path forward for now is to open a bug to get this fixed in the pipeline for future crawls, overwrite all `firstHtml` values in the `almanac` dataset to match `_is_base_page`, rerun all queries that include it, and update all results and written analyses that depend on those queries.

Thanks for pointing this out Paul!

[the remark this answers is not in the document]
Comment (rviscomi) [reply sql/2019/20_HTTP_2/20_04.sql]: I was about to file the bug and was querying the `summary_requests` dataset directly to eliminate any `almanac` dataset shenanigans, but it turns out that there are `almanac` dataset shenanigans!

The last query I shared in the previous comment produced hundreds of thousands of pages with more than 1 `firstHtml` request. It should have produced 0. This equivalent query of the `summary_requests` dataset _does_ produce 0 results:

```sql
#standardSQL
SELECT
  _TABLE_SUFFIX AS client,
  pageid,
  COUNT(0) AS html
FROM
  `httparchive.summary_requests.2019_07_01_*`
WHERE
  firstHtml
GROUP BY
  client,
  pageid
HAVING
  html > 1
ORDER BY
  html DESC
```

Now I think I know what the bug is. When I generated the almanac dataset, if the summary request had `firstHtml` all requests for that page/url were also annotated as `firstHtml`. I should be able to fix this, but we still need to rerun all the queries.

[the remark this answers is not in the document]
Comment (rviscomi) [reply sql/2019/20_HTTP_2/20_04.sql]: @paulcalvano for now let's change these queries to use `firstHtml` solely for consistency. At least it'll have the same bug as all the other metrics and we can fix them all in one swoop.

**Comment by rviscomi** [comment]
I've checked off all remaining metrics in #101 that are covered by this PR. Once the queries are updated to use `firstHtml` rather than `is_base_page` this should be ready for a final review to turn over to @bazzadp for writing.

**Comment by bazzadp** [comment]
@paulcalvano Can we change 20.16 to SUM rather than AVG?

```
  ROUND(SUM(num_requests),2) AS num_pushed_requests,
  ROUND(SUM(kb_transfered),2) AS num_kb_transfered
```

Also is there a way to group these into the below categories? I've done a vlookup in the results spreadsheet, but not sure if there is a better way for future years? And categorising by these categories is why I want the sum, rather than the total. I can approximate this (pretty well it turns out!) but multiplying the num pages by average but would be better to have the raw stats. I got VERY different results when I just looked at the averages...

- Video
- --
- Text
- PDF
- JavaScript
- Image
- Font
- Data
- CSS

**Comment by paulcalvano** [comment]
Hey Barry.  Sure, I'm updating the queries per Rick's comment earlier (ie, switching to the almanac table and using firstHtml instead of is_base_page). I'm sorry that I'm so behind on this.  

Will add a note to each sheet indicating whether the data is updated to use the almanac table.

**Comment by bazzadp** [comment]
Can I also get HTTPS added to the upgrade header spreadsheet?

**Comment by paulcalvano** [comment]
> Can I also get HTTPS added to the upgrade header spreadsheet?

Which query is that?

**Comment by paulcalvano** [comment]
I updated all the data in the sheet for the queries I updated above.  Will have the rest of these done later today.   THe numbers are very close - so the discrepancy I noted above does not appear to be too impactful and shouldn't affect the analysis.

**Comment by bazzadp** [comment]
>> Can I also get HTTPS added to the upgrade header spreadsheet?
> Which query is that?

The 4a, 5a and 6a sheet. Find that one most useful btw so ignore the individual queries as only going to look at this combined one.

**Comment by paulcalvano** [comment]
All queries have been updated to use the almanac tables except for 20.15.  There doesn't appear to be a summary_pages almanac table, so I left that one as is.   I've updated all the data in the sheets for the queries I updated.   In most cases, it resulted in only slight changes in the numbers. 

For 20.11, I included both tables because the results were very different between the two queries.    I trust the 2019_07_01 table results more FWIW - so we might just want to keep the only query for that one.   @rviscomi  wdyt?

**Comment by paulcalvano** [comment]
Just added the https indicators to two queries @bazzadp .  I didn't want to mess up your pivot tables, so instead of updating those in the sheet I added two extra sheets at the end with the results.

@rviscomi - this is ready for a final review.

**Pull request merged**